File: Common/Utilities.Serialization.vb
Imports System.Runtime.CompilerServices
Imports System.Text
Imports Media = System.Windows.Media
Imports Reflect = System.Reflection
Imports SIO = System.IO

Namespace Common

    Partial Class Utilities

        Public Class Serialization

            Private Sub New()
                Throw New NotSupportedException()
            End Sub

            Public Shared Function HexToColor(ByVal Hex As String) As Media.Color
                If Hex.StartsWith("#") Then
                    Hex = Hex.Substring(1)
                End If
                If Hex.Length = 8 Then
                    Return Media.Color.FromArgb(Byte.Parse(Hex.Substring(0, 2), Globalization.NumberStyles.HexNumber),
                                                Byte.Parse(Hex.Substring(2, 2), Globalization.NumberStyles.HexNumber),
                                                Byte.Parse(Hex.Substring(4, 2), Globalization.NumberStyles.HexNumber),
                                                Byte.Parse(Hex.Substring(6, 2), Globalization.NumberStyles.HexNumber))
                End If
                If Hex.Length = 6 Then
                    Return Media.Color.FromRgb(Byte.Parse(Hex.Substring(0, 2), Globalization.NumberStyles.HexNumber),
                                               Byte.Parse(Hex.Substring(2, 2), Globalization.NumberStyles.HexNumber),
                                               Byte.Parse(Hex.Substring(4, 2), Globalization.NumberStyles.HexNumber))
                End If

                Throw New ArgumentException("Invalid hex color.")
            End Function

            Public Shared Function ColorToHex(ByVal Color As Media.Color) As String
                Return String.Concat("#",
                                     Color.A.ToString("X2"),
                                     Color.R.ToString("X2"),
                                     Color.G.ToString("X2"),
                                     Color.B.ToString("X2"))
            End Function

            Private Shared Function EscapeChar(ByVal Ch As Char, ByVal EscapeChars As String) As String
                Select Case Ch
                    Case ControlChars.Cr
                        Return "\r"
                    Case ControlChars.Lf
                        Return "\n"
                    Case Else
                        If EscapeChars.Contains(Ch) Then
                            Return "\"c & Ch
                        End If
                        Return Ch
                End Select
            End Function

            Private Shared Function UnescapeChar(ByVal Ch As Char, ByVal EscapeChars As String) As Char
                Select Case Ch
                    Case "r"c
                        Return ControlChars.Cr
                    Case "n"c
                        Return ControlChars.Lf
                    Case Else
                        If EscapeChars.Contains(Ch) Then
                            Return Ch
                        End If
                        Throw New Exception("Invalid escape character.")
                End Select
            End Function

            Public Shared Function ListToString(ByVal List As IEnumerable(Of String)) As String
                Dim Res = New StringBuilder("{")
                Dim Bl = True

                For Each Str As String In List
                    If Bl Then
                        Bl = False
                    Else
                        Res.Append(",")
                    End If

                    For Each Ch In Str
                        Res.Append(EscapeChar(Ch, ",\{}"))
                    Next
                Next

                Return Res.Append("}").ToString()
            End Function

            Public Shared Function ListFromString(ByVal Str As String) As List(Of String)
                Dim Res = New List(Of String)()
                Dim R = New StringBuilder()

                For I As Integer = 0 To Str.Length - 1
                    Dim Ch = Str.Chars(I)

                    If I = 0 Then
                        If Ch <> "{"c Then
                            Throw New Exception("Invalid list string.")
                        End If
                        Continue For
                    End If

                    If Ch = "\"c Then
                        I += 1
                        If I = Str.Length Then
                            Throw New Exception("Invalid list string.")
                        End If

                        R.Append(UnescapeChar(Str.Chars(I), ",\{}"))
                        Continue For
                    End If

                    If Ch = ","c Or Ch = "}"c Then
                        Res.Add(R.ToString())
                        R.Clear()

                        If Ch = "}"c And I <> Str.Length - 1 Then
                            Throw New Exception("Invalid list string.")
                        End If

                        Continue For
                    End If

                    R.Append(Ch)
                Next

                Return Res
            End Function

            Public Shared Function ListToStringMultiline(ByVal List As IEnumerable(Of String)) As String
                Dim Res = New StringBuilder()

                For Each Str As String In List
                    For Each Ch In Str
                        Res.Append(EscapeChar(Ch, "\"))
                    Next

                    Res.AppendLine()
                Next

                Return Res.ToString()
            End Function

            Public Shared Function ListFromStringMultiline(ByVal Str As String) As List(Of String)
                Dim Res = New List(Of String)()
                Dim R = New StringBuilder()

                For I As Integer = 0 To Str.Length - 1
                    Dim Ch = Str.Chars(I)

                    If Ch = "\"c Then
                        I += 1
                        If I = Str.Length Then
                            Verify.Fail("Invalid list string.")
                        End If

                        R.Append(UnescapeChar(Str.Chars(I), "\"))
                        Continue For
                    End If

                    If Ch = ControlChars.Cr Or Ch = ControlChars.Lf Then
                        If (Ch = ControlChars.Cr And I + 1 < Str.Length) AndAlso Str.Chars(I + 1) = ControlChars.Lf Then
                            I += 1
                        End If

                        Res.Add(R.ToString())
                        R.Clear()

                        If I + 1 = Str.Length Then
                            Return Res
                        End If

                        Continue For
                    End If

                    R.Append(Ch)
                Next

                If Str.Length = 0 Then
                    Return Res
                End If

                Verify.Fail("Invalid list string.")
                Return Nothing
            End Function

            Public Shared Function DicToString(ByVal Dic As IDictionary(Of String, String)) As String
                Dim Res = New StringBuilder("{")
                Dim Bl = True

                For Each KV In Dic
                    If Bl Then
                        Bl = False
                    Else
                        Res.Append(",")
                    End If

                    For Each Ch In KV.Key
                        Res.Append(EscapeChar(Ch, ",\{}:"))
                    Next
                    Res.Append(":"c)
                    For Each Ch In KV.Value
                        Res.Append(EscapeChar(Ch, ",\{}"))
                    Next
                Next

                Return Res.Append("}").ToString()
            End Function

            Public Shared Function DicFromString(ByVal Str As String) As OrderedDictionary(Of String, String)
                Dim Res = New OrderedDictionary(Of String, String)()
                Dim R = New StringBuilder()
                Dim Key As String = Nothing

                For I As Integer = 0 To Str.Length - 1
                    Dim Ch = Str.Chars(I)

                    If I = 0 Then
                        If Ch <> "{"c Then
                            Throw New Exception("Invalid dictionary string.")
                        End If
                        Continue For
                    End If

                    If Ch = "\"c Then
                        I += 1
                        If I = Str.Length Then
                            Throw New Exception("Invalid dictionary string.")
                        End If

                        R.Append(UnescapeChar(Str.Chars(I), ",\{}:"))
                        Continue For
                    End If

                    If Ch = ":"c Then
                        Key = R.ToString()
                        R.Clear()

                        Continue For
                    End If

                    If Ch = ","c Or Ch = "}"c Then
                        If Key Is Nothing Then
                            Throw New Exception("Invalid dictionary string.")
                        End If
                        Res.Add(Key, R.ToString())
                        R.Clear()
                        Key = Nothing

                        If Ch = "}"c And I <> Str.Length - 1 Then
                            Throw New Exception("Invalid dictionary string.")
                        End If

                        Continue For
                    End If

                    R.Append(Ch)
                Next

                Return Res
            End Function

            Public Shared Function DicToStringMultiline(ByVal Dic As IDictionary(Of String, String)) As String
                Dim Res = New StringBuilder()

                For Each KV In Dic
                    For Each Ch In KV.Key
                        Res.Append(EscapeChar(Ch, "\:"))
                    Next
                    Res.Append(":"c)
                    For Each Ch In KV.Value
                        Res.Append(EscapeChar(Ch, "\"))
                    Next

                    Res.AppendLine()
                Next

                Return Res.ToString()
            End Function

            Public Shared Function DicFromStringMultiline(ByVal Str As String) As OrderedDictionary(Of String, String)
                Dim Res = New OrderedDictionary(Of String, String)()
                Dim R = New StringBuilder()
                Dim Key As String = Nothing

                For I As Integer = 0 To Str.Length - 1
                    Dim Ch = Str.Chars(I)

                    If Ch = "\"c Then
                        I += 1
                        Verify.False(I = Str.Length, "Invalid dictionary string.")

                        R.Append(UnescapeChar(Str.Chars(I), "\:"))
                        Continue For
                    End If

                    If Ch = ":"c And Key Is Nothing Then
                        Key = R.ToString()
                        R.Clear()

                        Continue For
                    End If

                    If Ch = ControlChars.Cr Or Ch = ControlChars.Lf Then
                        If (Ch = ControlChars.Cr And I + 1 < Str.Length) AndAlso Str.Chars(I + 1) = ControlChars.Lf Then
                            I += 1
                        End If

                        Verify.False(Key Is Nothing, "Invalid dictionary string.")
                        Res.Add(Key, R.ToString())
                        R.Clear()
                        Key = Nothing

                        If I + 1 = Str.Length Then
                            Return Res
                        End If

                        Continue For
                    End If

                    R.Append(Ch)
                Next

                Verify.Fail("Invalid dictionary string.")
                Return Nothing
            End Function

        End Class

    End Class

End Namespace

File: Common/Utilities.Serialization.cs
using System.Collections.Generic;
using System;
using System.Text;
using Media = System.Windows.Media;
using Reflect = System.Reflection;
using SIO = System.IO;

namespace Ks.Common
{
    partial class Utilities
    {
        public static class Serialization
        {

            public static Media.Color HexToColor(string Hex)
            {
                if (Hex.StartsWith("#"))
                    Hex = Hex.Substring(1);
                if (Hex.Length == 8)
                    return System.Windows.Media.Color.FromArgb(byte.Parse(Hex.Substring(0, 2), System.Globalization.NumberStyles.HexNumber), byte.Parse(Hex.Substring(2, 2), System.Globalization.NumberStyles.HexNumber), byte.Parse(Hex.Substring(4, 2), System.Globalization.NumberStyles.HexNumber), byte.Parse(Hex.Substring(6, 2), System.Globalization.NumberStyles.HexNumber));
                if (Hex.Length == 6)
                    return System.Windows.Media.Color.FromRgb(byte.Parse(Hex.Substring(0, 2), System.Globalization.NumberStyles.HexNumber), byte.Parse(Hex.Substring(2, 2), System.Globalization.NumberStyles.HexNumber), byte.Parse(Hex.Substring(4, 2), System.Globalization.NumberStyles.HexNumber));

                throw new ArgumentException("Invalid hex color.");
            }

            public static string ColorToHex(Media.Color Color)
            {
                return string.Concat("#", Color.A.ToString("X2"), Color.R.ToString("X2"), Color.G.ToString("X2"), Color.B.ToString("X2"));
            }

            private static string EscapeChar(char Ch, string EscapeChars)
            {
                switch (Ch)
                {
                    case '\r':
                        return @"\r";
                    case '\n':
                        return @"\n";
                    default:
                        var s = Ch.ToString();
                        if (EscapeChars.Contains(s))
                            return "\\" + s;
                        return s;
                }
            }

            private static char UnescapeChar(char Ch, string EscapeChars)
            {
                switch (Ch)
                {
                    case 'r':
                        return '\r';
                    case 'n':
                        return '\n';
                    default:
                        if (EscapeChars.Contains(Ch.ToString()))
                            return Ch;
                        throw new Exception("Invalid escape character.");
                }
            }

            public static string ListToString(IEnumerable<string> List)
            {
                var Res = new StringBuilder("{");

                foreach (var Str in List)
                {
                    foreach (var Ch in Str)
                        Res.Append(EscapeChar(Ch, @",\{}"));

                    // We always append a comma to the end of a list.
                    // This Is to distinguish {} from {""}.

                    Res.Append(",");
                }

                return Res.Append("}").ToString();
            }

            public static List<string> ListFromString(string Str)
            {
                var Res = new List<string>();
                var R = new StringBuilder();
                for (var I = 0; I < Str.Length; I++)
                {
                    var Ch = Str[I];

                    if (I == 0)
                    {
                        if (Ch != '{')
                            throw new Exception("Invalid list string.");
                        continue;
                    }

                    if (Ch == '\\')
                    {
                        I += 1;
                        if (I == Str.Length)
                            throw new Exception("Invalid list string.");

                        R.Append(UnescapeChar(Str[I], @",\{}"));
                        continue;
                    }

                    if ((Ch == ',') | (Ch == '}'))
                    {
                        // We have to ignore the last comma. See ListToString.
                        // But for backward compatibility, we add the non-empty strings to the list.
                        if (!((Ch == '}') & (R.Length == 0)))
                        {
                            Res.Add(R.ToString());
                            R.Clear();
                        }

                        if ((Ch == '}') & (I != (Str.Length - 1)))
                            throw new Exception("Invalid list string.");

                        continue;
                    }

                    R.Append(Ch);
                }

                return Res;
            }

            public static string ListToStringMultiline(IEnumerable<string> List)
            {
                var Res = new StringBuilder();

                foreach (var Str in List)
                {
                    foreach (var Ch in Str)
                        Res.Append(EscapeChar(Ch, @"\"));

                    Res.AppendLine();
                }

                return Res.ToString();
            }

            public static List<string> ListFromStringMultiline(string Str)
            {
                var Res = new List<string>();
                var R = new StringBuilder();
                for (var I = 0; I < Str.Length; I++)
                {
                    var Ch = Str[I];

                    if (Ch == '\\')
                    {
                        I += 1;
                        if (I == Str.Length)
                            Verify.Fail("Invalid list string.");

                        R.Append(UnescapeChar(Str[I], @"\"));
                        continue;
                    }

                    if ((Ch == '\r') | (Ch == '\n'))
                    {
                        if (((Ch == '\r') & ((I + 1) < Str.Length)) && Str[I + 1] == '\n')
                            I += 1;

                        Res.Add(R.ToString());
                        R.Clear();

                        if ((I + 1) == Str.Length)
                            return Res;

                        continue;
                    }

                    R.Append(Ch);
                }

                if (Str.Length == 0)
                    return Res;

                Verify.Fail("Invalid list string.");
                return null;
            }

            public static string DicToString(IDictionary<string, string> Dic)
            {
                var Res = new StringBuilder("{");
                var Bl = true;

                foreach (var KV in Dic)
                {
                    if (Bl)
                        Bl = false;
                    else
                        Res.Append(",");

                    foreach (var Ch in KV.Key)
                        Res.Append(EscapeChar(Ch, @",\{}:"));
                    Res.Append(':');
                    foreach (var Ch in KV.Value)
                        Res.Append(EscapeChar(Ch, @",\{}"));
                }

                return Res.Append("}").ToString();
            }

            public static OrderedDictionary<string, string> DicFromString(string Str)
            {
                var Res = new OrderedDictionary<string, string>();
                var R = new StringBuilder();
                string Key = null;
                for (var I = 0; I < Str.Length; I++)
                {
                    var Ch = Str[I];

                    if (I == 0)
                    {
                        if (Ch != '{')
                            throw new Exception("Invalid dictionary string.");
                        continue;
                    }

                    if (Ch == '\\')
                    {
                        I += 1;
                        if (I == Str.Length)
                            throw new Exception("Invalid dictionary string.");

                        R.Append(UnescapeChar(Str[I], @",\{}:"));
                        continue;
                    }

                    if (Ch == ':')
                    {
                        Key = R.ToString();
                        R.Clear();

                        continue;
                    }

                    if ((Ch == ',') | (Ch == '}'))
                    {
                        // If Key is nothing, then the collection must be empty.
                        Verify.True((Key == null).Implies((Res.Count == 0) & (R.Length == 0)), "Invalid dictionary string.");

                        if (Key != null)
                        {
                            Res.Add(Key, R.ToString());
                            R.Clear();
                            Key = null;
                        }

                        if ((Ch == '}') & (I != (Str.Length - 1)))
                            throw new Exception("Invalid dictionary string.");

                        continue;
                    }

                    R.Append(Ch);
                }

                return Res;
            }

            public static string DicToStringMultiline(IDictionary<string, string> Dic)
            {
                var Res = new StringBuilder();

                foreach (var KV in Dic)
                {
                    foreach (var Ch in KV.Key)
                        Res.Append(EscapeChar(Ch, @"\:"));
                    Res.Append(':');
                    foreach (var Ch in KV.Value)
                        Res.Append(EscapeChar(Ch, @"\"));

                    Res.AppendLine();
                }

                return Res.ToString();
            }

            public static OrderedDictionary<string, string> DicFromStringMultiline(string Str)
            {
                var Res = new OrderedDictionary<string, string>();
                var R = new StringBuilder();
                string Key = null;
                for (var I = 0; I < Str.Length; I++)
                {
                    var Ch = Str[I];

                    if (Ch == '\\')
                    {
                        I += 1;
                        Verify.False(I == Str.Length, "Invalid dictionary string.");

                        R.Append(UnescapeChar(Str[I], @"\:"));
                        continue;
                    }

                    if ((Ch == ':') & Key == null)
                    {
                        Key = R.ToString();
                        R.Clear();

                        continue;
                    }

                    if ((Ch == '\r') | (Ch == '\n'))
                    {
                        if (((Ch == '\r') & ((I + 1) < Str.Length)) && Str[I + 1] == '\n')
                            I += 1;

                        Verify.False(Key == null, "Invalid dictionary string.");
                        Res.Add(Key, R.ToString());
                        R.Clear();
                        Key = null;

                        if ((I + 1) == Str.Length)
                            return Res;

                        continue;
                    }

                    R.Append(Ch);
                }

                if (Str.Length == 0)
                    return Res;

                Verify.Fail("Invalid dictionary string.");
                return null;
            }
        }
    }
}
